transcom/mymove
# MB-2320 Prime API update shipment errors

## Description

This PR fixes the error messaging and some bugs with the Update MTO Shipment endpoint (`/mto-shipments/{:mtoShipmentID}`) endpoint in the Prime API.

Some of the issues addressed:
- Modifications to fields that can't be updated (such as Status, CustomerRemarks, ApprovedDate) being passed into the endpoint but a 200 response code being returned - the fields didn't update but the user wouldn't have known without comparing the output
- ScheduledPickupDate causing server errors
- No appropriate 404 response when a bad ID is passed in
- Unhelpful error messages for fields that didn't validate, no message, or the wrong response code
- Some unreachable blocks of code needed to be deleted

It adds two response codes to the endpoint:
- 409: Conflict Error - used when fields are passed into the request that cannot be updated by the prime
- 422: Validation Error - used when there are issues validating the input before update

The main updates are in `pkg/handlers/primeapi/mto_shipment.go` and `pkg/services/mto_shipment/mto_shipment_updater.go`.

## Setup

These changes can be tested with Postman or the Prime API Client.

### Testing with Postman

To set up Postman, please follow these instructions: https://github.com/transcom/mymove/wiki/setup-postman-to-make-mutual-tls-api-calls, and use these values in your environment:

- baseURL : `https://primelocal:9443/`
- prime : `prime/v1`
- support : `support/v1`

Once you have finished setting up your Postman environment, please follow these steps to test:

- Use the Fetch MTOs endpoint `{{baseUrl}}{{prime}}/move-task-orders` to select a shipment, making sure to get at minimum the shipment's ID and the eTag value
- Go to the Update MTO Shipment endpoint `{{baseUrl}}{{prime}}/mto-shipments/{:mtoShipmentID}` and enter the ID into the value for the `mtoShipmentID` in the path
- Next put the eTag value in the IfMatch Header
- Create a JSON Body including any of the fields you would like to try to update, such as:

```json
{
    "scheduledPickupDate": "2020-04-24",
    "actualPickupDate": "2020-04-26"
}
```

- Note the response and continue testing. At this point, if you have a successful update, you can grab the eTag from the response for this endpoint and should not have to go back to the Fetch MTOs endpoint (unless you forgot to update the eTag after you got the response).

### Testing with the CLI

Please follow these instructions: https://github.com/transcom/prime_api_deliverable/wiki/Prime-API-Client

These are the steps you should follow to test:

- Compile the binary using these commands:

`rm -f bin/prime-api-client`
`make bin/prime-api-client`

- Run `prime-api-client --insecure fetch-mtos | jq` and select a shipment from the output, noting the ID and eTag especially
- Create a JSON file with at minimum the ID of the shipment in it. Example:

```json
{
    "id": "<uuid>",
    "scheduledPickupDate": "2020-04-24",
    "actualPickupDate": "2020-04-26"
}
```

- Run `prime-api-client --insecure update-mto-shipment --etag {ETAG} --filename {PATH TO FILE} | jq` to test

NOTE: if there are any issues with the output from this command, drop the `| jq` and add `-v`. Some errors are not recognized as valid JSON by `jq`.

### Example Test Cases

- Try to update the `status`, which can only be updated by the TOO, or `requestedPickupDate`, which can only be updated by the customer. You should get a 409 response
- Try adding a MTO agent that doesn't exist on the original shipment. You should get a 422 response
- Try an ID that's not a shipment. You should get a 404

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2320) for this change
